webapp/sources/rudder/rudder-core/src/main/scala/com/normation/rudder/services/reports/ExecutionBatch.scala

```diff
@@ -939,25 +935,29 @@ final case class ContextForNoAnswer(
                                    u3 += t4-t3
 
                                    // okKeys is DirectiveId, ComponentName
-                                   val expected = okKeys.groupBy(_._1).map { case (directiveId, cptName) =>
-                                      DirectiveStatusReport(directiveId, cptName.toSeq.map { case (_, cpt) =>
-                                        val k = (directiveId, cpt)
-                                        val (policyMode, missingReportStatus, components) = expectedComponents(k)
-                                        (cpt, checkExpectedComponentWithReports(components, reports(k), missingReportStatus, policyMode, unexpectedInterpretation))
-                                      }.toMap)
-                                    }
+                                   val expected: Iterable[DirectiveStatusReport] =
+                                     expectedComponents.groupBy(_._1._1).map {
+                                       case (directiveId, expectedComponentsForDirective) =>
+                                         DirectiveStatusReport(directiveId, expectedComponentsForDirective.map {
+                                           case ((directiveId, components), (policyMode, missingReportStatus, component)) =>
+                                             val filteredReports = components.flatMap(c => reports.getOrElse((directiveId, c), Seq()))
+
+                                             (component.componentName, checkExpectedComponentWithReports(component, filteredReports, missingReportStatus, policyMode, unexpectedInterpretation))
+                                         })
+                                     }
+
 
                                    val t5 = System.nanoTime
                                    u4 += t5-t4
 
 
-                                   (missing, unexpected, expected)
+                                   ( unexpected, expected)
```

Comment:
I don't get why you don't have missing at all here: the case is still possible if some reports were lost (are more likely because of bad reporting logic somewhere)

Reply:
They are already treated  in another branch, i got duplicated missing if i keep it

Reply:
yes the difference is here: https://github.com/Normation/rudder/pull/3650/files/b5ee112bc450916897f1e3cbd6a032ff6feb5b21#diff-32bcd76fdbbd93efed1323a8e48c79320607d7ef97094e819936a74e6fab88feR938 

Previously we treated missing reports in its own place, so we were parsing all reports to check missing components so we produced our missing reports ,then when doing expected we were filtering missing key ( by using okKeys) 

in expected we now look for all expected keys, and when treating a component if we don't have reports we produce a missing

So we were going through all expected reports twice so find all missing and now this is done in one go